### PR TITLE
ddl, table: improve the efficiency of adding index

### PR DESCRIPTION
**Pull request by tangenta**
### What problem does this PR solve?

Issue Number: ref #47757

Problem Summary:

See #47757.

### What is changed and how it works?

This PR contains multiple optimizations:

- Change `MergeSortOverlapThreshold` to 8000 to increase the threshold to trigger merge-sort step.
- Optimize `extractDatumByOffsets` by decoding value into the datum directly instead of copying.
  ```
  BenchmarkExtractDatumByOffsets-10    	126919888	         9.298 ns/op
  BenchmarkExtractDatumByOffsets-10    	164564930	         7.231 ns/op
  ```
- Add index value buffer as a parameter to `GenIndexValue()` to avoid memory allocation.
- Change `table.IndexIter` from interface to struct so that it is not escaping to heap anymore.
  ```
   9601753               129.6 ns/op           138 B/op          4 allocs/op
  18007044                66.81 ns/op           17 B/op          2 allocs/op
  ```
- Adopt the fast path for `GenIndexKey` to reduce overhead.

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  >   

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: pkg/ddl/index.go [hunk 1]
```diff
@@ -1806,23 +1806,22 @@ func writeChunkToLocal(
...
 			return 0, nil, errors.Trace(err)
 		}
 		for i, index := range indexes {
 			idxID := index.Meta().ID
-			idxDataBuf = idxDataBuf[:0]
 			idxDataBuf = extractDatumByOffsets(
 				row, copCtx.IndexColumnOutputOffsets(idxID), c.ExprColumnInfos, idxDataBuf)
 			rsData := getRestoreData(c.TableInfo, copCtx.IndexInfo(idxID), c.PrimaryKeyInfo, handleDataBuf)
-			err = writeOneKVToLocal(ctx, writers[i], index, sCtx, writeBufs, idxDataBuf, rsData, handle)
+			err = writeOneKVToLocal(ctx, writers[i], index, sCtx, writeBufs, idxDataBuf, rsData, h)
```
Comment (wjhuang2016) [inline pkg/ddl/index.go:1818]: idxDataBuf[:len(copCtx.IndexColumnOutputOffsets(idxID))] ?

File: br/pkg/lightning/backend/external/iter.go [hunk 2]
```diff
@@ -86,8 +86,8 @@ type mergeIter[T heapElem, R sortedReader[T]] struct {
 	curr          T
 	lastReaderIdx int
 	err           error
-  
-  // determines whether to check reader hotspot, if hotspot is detected, we will
+
+	// determines whether to check reader hotspot, if hotspot is detected, we will
```
Comment (Benjamin2037) [inline br/pkg/lightning/backend/external/iter.go:90]: The comment seems not finished？

File: br/pkg/lightning/backend/external/writer.go [hunk 3]
```diff
@@ -47,7 +47,7 @@ var (
 
 	// MergeSortOverlapThreshold is the threshold of overlap between sorted kv files.
 	// if the overlap ratio is greater than this threshold, we will merge the files.
-	MergeSortOverlapThreshold int64 = 1000
+	MergeSortOverlapThreshold int64 = 8000
```
Comment (D3Hunter) [inline br/pkg/lightning/backend/external/writer.go:50]: for import, write&ingest step will adjust concurrency to read ingest-data depends on len(dataFiles). too many files might limit write speed.
maybe use another var on add-index
https://github.com/pingcap/tidb/blob/07fd1a433fd355880891779b7ec2a6e25f6829a0/br/pkg/lightning/backend/external/engine.go#L151

File: br/pkg/lightning/backend/external/iter.go (hunk 2 above)
Comment (tangenta) [reply br/pkg/lightning/backend/external/iter.go:90]: See next line.

File: br/pkg/lightning/backend/external/writer.go (hunk 3 above)
Comment (tangenta) [reply br/pkg/lightning/backend/external/writer.go]: 5b629ebb65e3d7af26eeeb750fbb33f535112e1e

File: br/pkg/lightning/backend/external/writer.go [hunk 4]
```diff
@@ -470,23 +439,9 @@ func (w *Writer) flushKVs(ctx context.Context, fromClose bool) (err error) {
 	}()
 
 	sortStart := time.Now()
-	if w.shareMu != nil {
-		sorty.MaxGor = min(8, uint64(variable.GetDDLReorgWorkerCounter()))
```
Comment (lance6716) [inline br/pkg/lightning/backend/external/writer.go:474]: we don't need `sorty` library? It has better performance

Comment (tangenta) [reply br/pkg/lightning/backend/external/writer.go:474]: Do you mean setting `sorty.MaxGor` to `1`?

Comment (tangenta) [reply br/pkg/lightning/backend/external/writer.go:474]: fd9caaa233a

**Review by wjhuang2016 — APPROVED** [review]

File: pkg/table/index.go [hunk 5]
```diff
@@ -86,15 +80,79 @@ type Index interface {
...
+	}
+	idxVal, err := iter.index.GenIndexValue(iter.sCtx, distinct, val, iter.handle, iter.handleRestoreData, valBuf)
+	if err != nil {
+		return nil, nil, false, err
+	}
+	iter.i++
+	return key, idxVal, distinct, err
+}
+
+// Valid returns true if the generator is exhausted.
```
Comment (lance6716) [inline pkg/table/index.go:152]:
```suggestion
// Valid returns true if the generator is not exhausted.
```

just a guess. In fact I don't know the meaning of `iter.i`

File: pkg/table/index.go [hunk 6]
```diff
@@ -86,15 +80,79 @@ type Index interface {
...
+func (iter *IndexKVGenerator) Valid() bool {
+	if len(iter.allIdxVals) == 0 {
+		return iter.i == 0
```
Comment (lance6716) [inline pkg/table/index.go:155]: it's only `Valid()` before the first `Next()` call?

**Review by lance6716 — COMMENTED** [review]
rest lgtm

File: pkg/table/index.go (hunk 6 above)
Comment (tangenta) [reply pkg/table/index.go:155]: Yes. For non multi-value indexes, there is only one index kv.

**Review by lance6716 — APPROVED** [review]

**Pull request merged**